### PR TITLE
[kubernetes_persistent_volume] fix azure pv disk kind

### PR DESCRIPTION
**Pull request by pchalamet**
### Description

This is an up-to-date fix for a long standing issue: inability to create a persistent volume from a managed disk on Azure.

### Acceptance tests
- [X] Have you added an acceptance test for the functionality being added?
- [X] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

### References

Original issue is here: https://github.com/terraform-providers/terraform-provider-kubernetes/issues/202

People attempted to provide fix but never reach the approved state:
* https://github.com/terraform-providers/terraform-provider-kubernetes/pull/342
* https://github.com/terraform-providers/terraform-provider-kubernetes/pull/586

So it's basically a merge from master from PR https://github.com/terraform-providers/terraform-provider-kubernetes/pull/586 with cleanup.
I've mostly done nothing - prior work from aforementioned PRs.

PR includes:
* Support for kubernetes_persistent_volume.spec.persistent_volume_source.azure_disk.kind
* Units tests
* Documentation of new feature
* No change to current behavior if kind is not provided

I've tested this on my production, this works perfectly and all volumes are now created this way.

Let me know if some more work is necessary. I will try to do my best to make this into master.

### CODE REVIEW

**Comment by pchalamet** [comment]
Hi,

Checks do not pass but that's an internal build error (I did not select "Allow edits and access to secrets by maintainers" on #836, but still same error with flag set).

Hope you can still review this PR.
Thanks!

File: kubernetes/resource_kubernetes_persistent_volume_test.go [hunk 1]
```diff
@@ -880,6 +905,53 @@ resource "aws_ebs_volume" "test" {
...
+			azure_disk {
+				caching_mode = "None"
+				data_disk_uri = "${azurerm_managed_disk.dbstorage.id}"
+				disk_name = "%s"
+				kind = "Managed"
+			}
+		}
+	}
+}
+resource "azurerm_managed_disk" "dbstorage" {
```
Comment (alexsomesan) [inline kubernetes/resource_kubernetes_persistent_volume_test.go:940]: Due to the way the testing framework is designed, you would have to explicitly include the AzureRM provider in the test setup here: https://github.com/terraform-providers/terraform-provider-kubernetes/blob/69a6932390c68a14052d1e732fce3799f78d5c5c/kubernetes/provider_test.go#L25

Use the AWS or Google entries already in there as a guideline.

Comment (pchalamet) [reply kubernetes/resource_kubernetes_persistent_volume_test.go:940]: Added:
``
go get github.com/terraform-providers/terraform-provider-azurerm/azurerm
``
and modified map as requested.

Comment (pchalamet) [reply kubernetes/resource_kubernetes_persistent_volume_test.go:940]: Hi Alex, is it ok ? Does this solve the issue ?

**Comment by pchalamet** [comment]
Sorry for the mess on this branch but I wanted to clean up things a little bit.
One things strange: go modules are versioned and this obviously generates a lot of noise on this PR. Feel free to revert 26ca833 if required - but this won't pass tests... Well, it's up to you now, I can't do much more now!
Thanks!

**Comment by pchalamet** [comment]
Hey @alexsomesan ,

Looks like it's pretty much stalled here :-(
What can be enhanced to make progress on this PR ?

Thanks !

**Comment by dak1n1** [comment]
@pchalamet hi! I just wanted to let you know that the PR had been queued up in our system for a while, but it finally reached the top of the list and I've started some testing on it. If you would like to do a rebase, that's one potential path forward. Or since I've already done the rebase on my local branch, I'd be happy to push that change into a different PR and work from there. Let me know how you would like to proceed. Thanks!

**Comment by pchalamet** [comment]
hi @dak1n1, glad to read this !
Looks like the branch is up-to-date (and mergeable into master), let me know if something has to be done.
But if you want to take over the PR, no problem - I just want this to land into master branch!

**Comment by dak1n1** [comment]
oh ok, I see that too now. Thanks for letting me know. Maybe I had it confused with something else... I'll double check my local branch. We can keep working in this PR then. :smile:

**Comment by aareet** [comment]
That's odd because I do see that the branch has conflicts with master...weird

Screen Shot 2020-08-24 at 2 59 32 PM

**Comment by pchalamet** [comment]
Yes that's pretty weird. I see everything looks good on my side.
Btw I remember I pushed force several times to fix hazardous go mod updates... maybe related.

If you prefer to continue on your own branch, there is no problem!

**Comment by dak1n1** [comment]
oh cool, someone else saw it too! That's ok. I'll make sure it doesn't need a rebase before I merge. Thanks for double checking that. Here's what mine looks like now: 
image

**Comment by pchalamet** [comment]
Hi @dak1n1 !
I'm just seeing the branch has conflicts due to merge into main branch.
Do you want me to resolve issues or are you taking care of this ?

Thanks.

**Comment by dak1n1** [comment]
@pchalamet I'm taking care of the dependencies in this PR and doing some upgrade refactoring with it.  https://github.com/hashicorp/terraform-provider-kubernetes/pull/968

After that merges, I'll open another PR to pull in your changes, just so I can make adjustments and test things without too much back-and-forth communication. Thanks for checking.

**Comment by pchalamet** [comment]
Hi @dak1n1 ,

I recently switched to Terraform 0.13 - which brings some interesting changes how plugins are discovered and unfortunately, even following the upgrade guide, I'm a bit stuck checking if working correctly on this new version.

I used to put the plugin into `~/.terraform.d/plugins` but looks like it's over now. So I put it into `~/.terraform.d/plugins/registry.terraform.io/hashicorp/kubernetes/0.0.0/darwin_amd64`, then tf find the local folder but just decide not to use it and fail with an error.

I must do something wrong here obviously :-(
Would you mind letting me know how you test this plugin straight from the local build using tf 0.13 ? Thanks!

**Comment by alexsomesan** [comment]
> Hi @dak1n1 ,
> 
> I recently switched to Terraform 0.13 - which brings some interesting changes how plugins are discovered and unfortunately, even following the upgrade guide, I'm a bit stuck checking if working correctly on this new version.
> 
> I used to put the plugin into `~/.terraform.d/plugins` but looks like it's over now. So I put it into `~/.terraform.d/plugins/registry.terraform.io/hashicorp/kubernetes/0.0.0/darwin_amd64`, then tf find the local folder but just decide not to use it and fail with an error.
> 
> I must do something wrong here obviously :-(
> Would you mind letting me know how you test this plugin straight from the local build using tf 0.13 ? Thanks!

Version `0.0.0` won't work, unfortunately. You have to set the version to something like `9.9.9` for example, to also avoid conflicts with actual releases. Also add version to the provider binary filename (e.g. `terraform-provider-kubernetes_9.9.9_darwin_amd64`).

**Comment by pchalamet** [comment]
Thanks for the info @alexsomesan, will give it a try.

**Comment by dak1n1** [comment]
@pchalamet  I cherry-picked your changes into the PR here: https://github.com/hashicorp/terraform-provider-kubernetes/pull/1011

Thanks for all your work on this!

**Comment by pchalamet** [comment]
Thanks a lot @dak1n1 for the last mile!

**Pull request closed without merging**